File: skywalking/java/src/main/java/com/sagarnikam123/FuzzyTrainSkywalking.java
package com.sagarnikam123;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetAddress;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.Random;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Fake log generator with SkyWalking integration (Java version).
 * 
 * Based on fuzzy-train.py, generates fake logs and sends them to SkyWalking 
 * via the Java agent's gRPC log reporter (logback toolkit).
 * 
 * Logs appear in SkyWalking UI under General Service → <service-name> → Log tab.
 * 
 * Requires SkyWalking Java Agent with logback toolkit configured.
 * 
 * Usage:
 *   java -javaagent:/path/to/skywalking-agent.jar -jar fuzzy-train-skywalking.jar
 *   java -jar fuzzy-train-skywalking.jar --lines-per-second 5
 */
public class FuzzyTrainSkywalking {
    
    private static final Logger logger = LoggerFactory.getLogger(FuzzyTrainSkywalking.class);
    private static final Random random = new Random();
    private static final AtomicBoolean running = new AtomicBoolean(true);
    private static final AtomicLong logCount = new AtomicLong(0);
    private static final AtomicLong traceIdCounter = new AtomicLong(1);
    
    private static final String VERSION = "1.0.0";
    private static final double DETAIL_PROBABILITY = 0.3;
    
    // Log levels (matching fuzzy-train.py)
    private static final List<String> LOG_LEVELS = List.of("INFO", "ERROR", "DEBUG", "WARN");
    
    // Sentences (matching fuzzy-train.py)
    private static final List<String> SENTENCES = List.of(
        "Processing request from client.",
        "Database connection established successfully.",
        "Cache hit ratio is below threshold.",
        "User authentication completed.",
        "API request processing time exceeded limits.",
        "Memory usage is within normal parameters.",
        "Disk I/O operations completed.",
        "Network latency detected on primary interface.",
        "Configuration loaded from environment variables.",
        "Background task scheduler initiated.",
        "Garbage collection cycle completed.",
        "Service health check passed.",
        "Rate limiting applied to incoming requests.",
        "Thread pool resources allocated.",
        "Security policy validation completed.",
        "Data synchronization process started.",
        "Backup procedure executed successfully.",
        "Input validation performed on user data.",
        "Rendering engine initialized with default parameters.",
        "Encryption key rotation completed."
    );
    
    // Default configuration constants
    private static final int DEFAULT_MIN_LOG_LENGTH = 90;
    private static final int DEFAULT_MAX_LOG_LENGTH = 100;
    private static final double DEFAULT_LINES_PER_SECOND = 1.0;
    private static final String DEFAULT_TRACE_ID_TYPE = "pid";
    
    private static String processId;
    
    /**
     * Get process identifier based on environment (PID for local, container ID for containers).
     */
    private static String getProcessId() {
        // Check if running in container
        boolean inContainer = Files.exists(Paths.get("/.dockerenv")) ||
                              Files.exists(Paths.get("/proc/1/cgroup")) ||
                              System.getenv("container") != null ||
                              Files.exists(Paths.get("/run/.containerenv"));
        
        if (inContainer) {
            try {
                String hostname = InetAddress.getLocalHost().getHostName();
                // For Kubernetes pods, extract the hash suffix
                if (hostname.contains("-")) {
                    String[] parts = hostname.split("-");
                    if (parts.length >= 2) {
                        String suffix = parts.length > 2 
                            ? parts[parts.length - 2] + "-" + parts[parts.length - 1]
                            : parts[parts.length - 1];
                        return suffix.length() > 12 ? suffix.substring(0, 12) : suffix;
                    }
                }
                return hostname.length() > 12 ? hostname.substring(0, 12) : hostname;
            } catch (Exception e) {
                return "unknown";
            }
        } else {
            return String.valueOf(ProcessHandle.current().pid());
        }
    }
    
    /**
     * Generate a random log message of specified length.
     */
    private static String generateRandomMessage(int length) {
        StringBuilder message = new StringBuilder();
        while (message.length() < length) {
            String sentence = SENTENCES.get(random.nextInt(SENTENCES.size()));
            if (random.nextDouble() < DETAIL_PROBABILITY) {
                String detail = generateRandomString(random.nextInt(21) + 10);
                sentence += " Details: " + detail;
            }
            message.append(sentence).append(" ");
        }
        return message.substring(0, Math.min(message.length(), length));
    }
    
    /**
     * Generate random alphanumeric string.
     */
    private static String generateRandomString(int length) {
        String chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        StringBuilder sb = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            sb.append(chars.charAt(random.nextInt(chars.length())));
        }
        return sb.toString();
    }
    
    /**
     * Generate trace ID for log correlation.
     */
    private static String generateTraceId(boolean includeTraceId, String traceIdType) {
        if (!includeTraceId) {
            return null;
        }
        
        long counter = traceIdCounter.getAndIncrement();
        if ("pid".equals(traceIdType)) {
            return String.format("%s-%08d", processId, counter);
        } else {
            return String.format("%08d", counter);
        }
    }
    
    /**
     * Log message at specified level.
     */
    private static void logAtLevel(String level, String message) {
        switch (level) {
            case "INFO" -> logger.info(message);
            case "WARN" -> logger.warn(message);
            case "ERROR" -> logger.error(message);
            case "DEBUG" -> logger.debug(message);
        }
    }
    
    public static void main(String[] args) {
        // Initialize process ID
        processId = getProcessId();
        
        // Parse arguments
        double linesPerSecond = DEFAULT_LINES_PER_SECOND;
        int minLogLength = DEFAULT_MIN_LOG_LENGTH;
        int maxLogLength = DEFAULT_MAX_LOG_LENGTH;
        boolean includeTraceId = true;
        String traceIdType = DEFAULT_TRACE_ID_TYPE;
        
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--lines-per-second", "-l" -> {
                    if (i + 1 < args.length) {
                        linesPerSecond = Double.parseDouble(args[++i]);
                    }
                }
                case "--min-log-length" -> {
                    if (i + 1 < args.length) {
                        minLogLength = Integer.parseInt(args[++i]);
                    }
                }
                case "--max-log-length" -> {
                    if (i + 1 < args.length) {
                        maxLogLength = Integer.parseInt(args[++i]);
                    }
                }
                case "--no-trace-id" -> includeTraceId = false;
                case "--trace-id-type" -> {
                    if (i + 1 < args.length) {
                        traceIdType = args[++i].toLowerCase();
                    }
                }
                case "--help", "-h" -> {
                    printHelp();
                    return;
                }
                case "--version", "-v" -> {
                    System.out.println("fuzzy-train-skywalking-java " + VERSION);
                    return;
                }
            }
        }
        
        // Validate length params (matching fuzzy-train.py logic)
        if (minLogLength != DEFAULT_MIN_LOG_LENGTH && maxLogLength == DEFAULT_MAX_LOG_LENGTH) {
            maxLogLength = minLogLength;
        } else if (minLogLength == DEFAULT_MIN_LOG_LENGTH && maxLogLength != DEFAULT_MAX_LOG_LENGTH) {
            minLogLength = maxLogLength;
        }
        
        if (minLogLength > maxLogLength) {
            logger.error("min-log-length ({}) cannot be greater than max-log-length ({})", minLogLength, maxLogLength);
            System.exit(1);
        }
        
        // Register shutdown hook
        final long[] finalCount = {0};
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            running.set(false);
            logger.info("Shutdown initiated. Generated {} total log entries.", logCount.get());
        }));
        
        long intervalMs = (long) (1000.0 / linesPerSecond);
        
        String serviceName = System.getenv().getOrDefault("SW_AGENT_NAME", "skywalking-test::fuzzy-train-java");
        String backend = System.getenv().getOrDefault("SW_AGENT_COLLECTOR_BACKEND_SERVICES", "skywalking-satellite.skywalking.svc:11800");
        
        logger.info("Starting fuzzy-train-skywalking-java v{}", VERSION);
        logger.info("Service: {}", serviceName);
        logger.info("Backend: {}", backend);
        logger.info("Rate: {} logs/second", linesPerSecond);
        logger.info("Message length: {}-{} chars", minLogLength, maxLogLength);
        logger.info("Trace ID: {}", includeTraceId ? "enabled (" + traceIdType + ")" : "disabled");
        
        final int finalMinLen = minLogLength;
        final int finalMaxLen = maxLogLength;
        final boolean finalIncludeTraceId = includeTraceId;
        final String finalTraceIdType = traceIdType;
        
        while (running.get()) {
            try {
                String level = LOG_LEVELS.get(random.nextInt(LOG_LEVELS.size()));
                String traceId = generateTraceId(finalIncludeTraceId, finalTraceIdType);
                int messageLength = random.nextInt(finalMaxLen - finalMinLen + 1) + finalMinLen;
                String message = generateRandomMessage(messageLength);
                
                // Add trace_id to message if enabled (for visibility in logs)
                String fullMessage = traceId != null ? "[" + traceId + "] " + message : message;
                
                logAtLevel(level, fullMessage);
                
                long count = logCount.incrementAndGet();
                if (count % 100 == 0) {
                    logger.info("Generated {} log entries", count);
                }
                
                Thread.sleep(intervalMs);
                
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            } catch (Exception e) {
                logger.error("Error generating log: {}", e.getMessage());
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
        
        logger.info("Shutdown complete. Generated {} total log entries.", logCount.get());
    }
    
    private static void printHelp() {
        System.out.println("""
            fuzzy-train-skywalking-java - Fake log generator with SkyWalking integration
            
            Based on fuzzy-train.py, generates fake logs and sends them to SkyWalking
            via the Java agent's gRPC log reporter (logback toolkit).
            
            Usage:
              java -javaagent:/sky/agent/skywalking-agent.jar -jar fuzzy-train-skywalking.jar [options]
            
            Basic Options:
              -l, --lines-per-second <n>  Generation rate (default: 1)
              -h, --help                  Show this help message
              -v, --version               Show version
            
            Log Content:
              --min-log-length <n>        Minimum message length in characters (default: 90)
              --max-log-length <n>        Maximum message length in characters (default: 100)
            
            Field Control:
              --no-trace-id               Exclude trace_id field
              --trace-id-type <type>      Trace ID type: pid or integer (default: pid)
            
            Environment Variables:
              SW_AGENT_NAME                         Service name in SkyWalking
              SW_AGENT_COLLECTOR_BACKEND_SERVICES   Satellite/OAP endpoint
              SW_GRPC_LOG_SERVER_HOST               Log server host
              SW_GRPC_LOG_SERVER_PORT               Log server port
            
            Examples:
              # Default: 1 log per second
              java -javaagent:/sky/agent/skywalking-agent.jar -jar fuzzy-train-skywalking.jar
              
              # High volume: 10 logs per second
              java -javaagent:/sky/agent/skywalking-agent.jar -jar fuzzy-train-skywalking.jar -l 10
              
              # Custom message length
              java -javaagent:/sky/agent/skywalking-agent.jar -jar fuzzy-train-skywalking.jar \\
                --min-log-length 150 --max-log-length 200
            """);
    }
}

File: skywalking/python/fuzzy-train-skywalking.py
#!/usr/bin/env python3
"""
fuzzy-train-skywalking.py - Fake log generator with SkyWalking integration

Based on fuzzy-train.py, this version sends logs to SkyWalking via the Python agent's log reporter.
Logs appear in SkyWalking UI under General Service → <service-name> → Log tab.

Usage:
    python3 fuzzy-train-skywalking.py --help
    python3 fuzzy-train-skywalking.py --lines-per-second 2

Environment Variables:
    SW_AGENT_NAME                         - Service name (default: skywalking-test::fuzzy-train)
    SW_AGENT_COLLECTOR_BACKEND_SERVICES   - Satellite/OAP endpoint (default: skywalking-satellite.skywalking.svc:11800)
    SW_AGENT_LOG_REPORTER_ACTIVE          - Enable log reporter (default: true)
    SW_AGENT_DISABLE                      - Set to 'true' to disable agent (for local testing)
"""

import argparse
import logging
import os
import random
import signal
import socket
import string
import sys
import time
from datetime import datetime, timezone
from typing import Optional

# SkyWalking agent configuration
SW_SERVICE_NAME = os.getenv('SW_AGENT_NAME', 'skywalking-test::fuzzy-train')
SW_BACKEND = os.getenv('SW_AGENT_COLLECTOR_BACKEND_SERVICES', 'skywalking-satellite.skywalking.svc:11800')
SW_AGENT_ENABLED = os.getenv('SW_AGENT_DISABLE', 'false').lower() != 'true'

# Try to initialize SkyWalking agent (may fail locally without proper setup)
if SW_AGENT_ENABLED:
    try:
        from skywalking import agent, config
        config.init(
            agent_collector_backend_services=SW_BACKEND,
            agent_name=SW_SERVICE_NAME,
            agent_log_reporter_active=True,
            agent_log_reporter_level='DEBUG',
            agent_logging_level='WARNING',  # Agent's own logging (not app logs)
        )
        agent.start()
        print(f"[INFO] SkyWalking agent started: {SW_SERVICE_NAME} -> {SW_BACKEND}", file=sys.stderr)
    except ImportError as e:
        print(f"[WARN] SkyWalking agent not available: {e}. Running without agent.", file=sys.stderr)
        SW_AGENT_ENABLED = False
    except Exception as e:
        print(f"[WARN] SkyWalking agent failed to start: {e}. Running without agent.", file=sys.stderr)
        SW_AGENT_ENABLED = False
else:
    print("[INFO] SkyWalking agent disabled (SW_AGENT_DISABLE=true)", file=sys.stderr)

# Now set up application logging
logger = logging.getLogger('fuzzy-train')
logger.setLevel(logging.DEBUG)

# Console handler for stdout
console_handler = logging.StreamHandler(sys.stdout)
console_handler.setLevel(logging.DEBUG)
console_formatter = logging.Formatter(
    '%(asctime)s [%(levelname)s] %(name)s - %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)
console_handler.setFormatter(console_formatter)
logger.addHandler(console_handler)

# Constants (matching fuzzy-train.py)
__version__ = "1.0.0"
DETAIL_PROBABILITY = 0.3
TRACE_ID_COUNTER = 1

LOG_LEVELS = ["INFO", "ERROR", "DEBUG", "WARN"]

SENTENCES = [
    "Processing request from client.",
    "Database connection established successfully.",
    "Cache hit ratio is below threshold.",
    "User authentication completed.",
    "API request processing time exceeded limits.",
    "Memory usage is within normal parameters.",
    "Disk I/O operations completed.",
    "Network latency detected on primary interface.",
    "Configuration loaded from environment variables.",
    "Background task scheduler initiated.",
    "Garbage collection cycle completed.",
    "Service health check passed.",
    "Rate limiting applied to incoming requests.",
    "Thread pool resources allocated.",
    "Security policy validation completed.",
    "Data synchronization process started.",
    "Backup procedure executed successfully.",
    "Input validation performed on user data.",
    "Rendering engine initialized with default parameters.",
    "Encryption key rotation completed."
]

# Default configuration constants
DEFAULT_MIN_LOG_LENGTH = 90
DEFAULT_MAX_LOG_LENGTH = 100
DEFAULT_LINES_PER_SECOND = 1
DEFAULT_TRACE_ID_TYPE = "pid"
DEFAULT_TIME_ZONE = "local"

def get_process_id() -> str:
    """Get process identifier based on environment (PID for local, container ID for containers)."""
    # Check if running in any container
    if (os.path.exists('/.dockerenv') or
        os.path.exists('/proc/1/cgroup') or
        os.environ.get('container') or
        os.path.exists('/run/.containerenv')):  # Podman

        hostname = socket.gethostname()

        # For Kubernetes pods, extract the hash suffix
        if '-' in hostname:
            parts = hostname.split('-')
            if len(parts) >= 2:
                return f"{parts[-2]}-{parts[-1]}"[:12] if len(parts) > 2 else parts[-1][:12]

        # Fallback to truncated hostname for Docker/Podman
        return hostname[:12]
    else:
        return str(os.getpid())

PID = get_process_id()

def generate_random_message(length: int) -> str:
    """Generate a random log message of specified length."""
    message = ""
    while len(message) < length:
        sentence = random.choice(SENTENCES)
        if random.random() < DETAIL_PROBABILITY:
            detail = ''.join(random.choices(string.ascii_letters + string.digits, k=random.randint(10, 30)))
            sentence += f" Details: {detail}"
        message += sentence + " "
    return message[:length]

def generate_timestamp(time_zone: str) -> str:
    """Generate ISO 8601 timestamp in specified timezone."""
    now = datetime.now(timezone.utc)
    if time_zone.lower() == "local":
        now = now.astimezone()
    if time_zone.upper() == "UTC":
        return now.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
    else:
        offset = now.strftime('%z')
        offset_fmt = f"{offset[:3]}:{offset[3:]}" if offset else ""
        return now.strftime(f"%Y-%m-%dT%H:%M:%S.%f{offset_fmt}")

def generate_trace_id(include_trace_id: bool, trace_id_type: str) -> Optional[str]:
    """Generate trace ID for log correlation."""
    global TRACE_ID_COUNTER
    if not include_trace_id:
        return None

    if trace_id_type == "pid":
        trace_id = f"{PID}-{TRACE_ID_COUNTER:08d}"
    else:  # integer
        trace_id = f"{TRACE_ID_COUNTER:08d}"

    TRACE_ID_COUNTER += 1
    return trace_id

def log_message(level: str, message: str, trace_id: Optional[str] = None):
    """Log a message at the specified level with optional trace_id in extra."""
    extra = {'trace_id': trace_id} if trace_id else {}
    
    if level == 'INFO':
        logger.info(message, extra=extra)
    elif level == 'WARN':
        logger.warning(message, extra=extra)
    elif level == 'ERROR':
        logger.error(message, extra=extra)
    elif level == 'DEBUG':
        logger.debug(message, extra=extra)

# Signal handler for graceful shutdown
running = True

def signal_handler(signum, frame):
    global running
    logger.info(f"Received signal {signum}, shutting down gracefully...")
    running = False

signal.signal(signal.SIGINT, signal_handler)
signal.signal(signal.SIGTERM, signal_handler)

def parse_args() -> argparse.Namespace:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description='Fake log generator with SkyWalking integration (based on fuzzy-train)',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Default: 1 log per second
  python3 fuzzy-train-skywalking.py

  # High volume: 10 logs per second
  python3 fuzzy-train-skywalking.py --lines-per-second 10

  # Custom message length
  python3 fuzzy-train-skywalking.py --min-log-length 150 --max-log-length 200

  # Without trace ID
  python3 fuzzy-train-skywalking.py --no-trace-id

Environment Variables:
  SW_AGENT_NAME                         Service name in SkyWalking
  SW_AGENT_COLLECTOR_BACKEND_SERVICES   Satellite/OAP endpoint
        """
    )
    
    parser.add_argument('-v', '--version', action='version', version=f'fuzzy-train-skywalking {__version__}')
    
    # Basic Options
    basic = parser.add_argument_group('Basic Options')
    basic.add_argument('--lines-per-second', type=float, default=DEFAULT_LINES_PER_SECOND,
                       help=f'Generation rate (default: {DEFAULT_LINES_PER_SECOND})')
    
    # Log Content
    content = parser.add_argument_group('Log Content')
    content.add_argument('--min-log-length', type=int, default=DEFAULT_MIN_LOG_LENGTH,
                         help=f'Minimum message length in characters (default: {DEFAULT_MIN_LOG_LENGTH})')
    content.add_argument('--max-log-length', type=int, default=DEFAULT_MAX_LOG_LENGTH,
                         help=f'Maximum message length in characters (default: {DEFAULT_MAX_LOG_LENGTH})')
    content.add_argument('--time-zone', type=str, default=DEFAULT_TIME_ZONE,
                         choices=['local', 'UTC', 'utc', 'LOCAL'],
                         help=f'Timestamp timezone (default: {DEFAULT_TIME_ZONE})')
    
    # Field Control
    fields = parser.add_argument_group('Field Control')
    fields.add_argument('--no-trace-id', action='store_true',
                        help='Exclude trace_id field')
    fields.add_argument('--trace-id-type', type=str, default=DEFAULT_TRACE_ID_TYPE,
                        choices=['pid', 'integer'],
                        help=f'Trace ID type: pid or integer (default: {DEFAULT_TRACE_ID_TYPE})')
    
    return parser.parse_args()

def main():
    args = parse_args()
    
    # Validate length params
    min_len = args.min_log_length
    max_len = args.max_log_length
    
    if min_len != DEFAULT_MIN_LOG_LENGTH and max_len == DEFAULT_MAX_LOG_LENGTH:
        max_len = min_len
    elif min_len == DEFAULT_MIN_LOG_LENGTH and max_len != DEFAULT_MAX_LOG_LENGTH:
        min_len = max_len
    
    if min_len > max_len:
        logger.error(f"min-log-length ({min_len}) cannot be greater than max-log-length ({max_len})")
        sys.exit(1)
    
    lps = args.lines_per_second
    include_trace_id = not args.no_trace_id
    trace_id_type = args.trace_id_type.lower()
    tz = args.time_zone
    
    interval = 1.0 / lps if lps > 0 else 1.0
    
    logger.info(f"Starting fuzzy-train-skywalking v{__version__}")
    logger.info(f"Service: {SW_SERVICE_NAME}")
    logger.info(f"Backend: {SW_BACKEND}")
    logger.info(f"Rate: {lps} logs/second")
    logger.info(f"Message length: {min_len}-{max_len} chars")
    logger.info(f"Trace ID: {'enabled (' + trace_id_type + ')' if include_trace_id else 'disabled'}")
    
    count = 0
    while running:
        try:
            log_level = random.choice(LOG_LEVELS)
            trace_id = generate_trace_id(include_trace_id, trace_id_type)
            message_length = random.randint(min_len, max_len)
            message = generate_random_message(message_length)
            
            # Add trace_id to message if enabled (for visibility in logs)
            if trace_id:
                full_message = f"[{trace_id}] {message}"
            else:
                full_message = message
            
            log_message(log_level, full_message, trace_id)
            
            count += 1
            if count % 100 == 0:
                logger.info(f"Generated {count} log entries")
            
            time.sleep(interval)
            
        except Exception as e:
            logger.error(f"Error generating log: {e}")
            time.sleep(1)
    
    logger.info(f"Shutdown complete. Generated {count} total log entries.")

if __name__ == '__main__':
    main()
